[fix] WebRTC 버그 수정 및 마이크, 카메라 on/off 관련 소켓 이벤트 추가

HyewonKkang opened the pull request:
## 상세내용
- WebRTC 관련해서 버그들을 수정했어요.
- remote 유저의 마이크와 카메라의 on/off 상태 변화를 전달하기 위한 소켓 이벤트를 추가했어요.
- 기존 userList를 userStreamList로 통일시켰어요.
- 추가로 남은 작업은 1) camera off인 경우 중앙에 이름 표시, 2) 카메라, 마이크 디바이스가 없는 사람 처리 가 있어요.

HyewonKkang committed:
fix: [FE] 비디오 디바이스 없는 유저 처리, dto 수정

- Intent

yscriuf committed:
feat: [BE] change-stream 소켓 이벤트 핸들러 추가
- Intent
특정 유저가 카메라, 마이크 상태를 변경할 때 모든 유저들에게 전달

yscriuf committed:
fix: [FE] change-stream 이벤트 내용 수정으로 인한 DTO 롤백
- Intent
기존 join-lobby에 담아서 보내던 마이크, 카메라 상태를 change-stream 이벤트에서 전달하도록 수정하였습니다.

yscriuf committed:
feat: [FE] change-stream 소켓 이벤트 추가

yscriuf committed:
feat: [FE, BE] user DTO에 video, audio 추가

- Intent
change-stream 소켓 이벤트 만으로는 기존에 접속해있던 유저의 stream 정보를 받아오지 못하는 문제가 있었어요.
join-lobby의 return 값에 접속해있던 모든 유저의 stream 정보를 담아서 전달하기로 했어요.

yscriuf committed:
refactor: [FE,BE] change-stream에서 update-user-stream으로 이벤트명 변경

HyewonKkang committed:
feat: [FE] remote 유저의 마이크/캠 상태 변경 반영

- Intent

yscriuf committed:
fix: [BE] leave-lobby의 payload 수정
-Intent
leave-game의 payload와 일치시키기 위함입니다.

HyewonKkang committed:
fix: [FE] userList 삭제 후 userStreamList로 통일

- Intent
기존에 join-lobby event마다 userList에 user 정보를 추가해주었는데
WebRTC를 구현하면서 stream 정보까지 포함해서 저장하는 방향으로
수정했어요.

HyewonKkang committed:
fix: [FE] 카메라 off로 로비 진입 시의 오류 수정

- getUserMedia() 요청을 보내는 부분에서 video와 audio를 모두 true로 변경
- CameraButton.tsx 오타 수정
- update-user-stream 소켓이벤트 off 추가

- Intent
유저 스트림의 video와 audio 스트림을 모두 요청한 후에 상태를 변경해야
해서 초기에 모두 true로 요청해요.

HyewonKkang committed:
fix: [FE] WebRTC offer 전달 시의 setTimeout() 제거

- Intent
기존에는 유저 스트림을 가져오기 전에 offer를 전달해 상대에게 내 비디오가
뜨지 않는 문제가 있었어요. 이를 메인에서 비디오 스트림을 만들고 로비로
들어와 offer를 보내는 식으로 수정했어요.

HyewonKkang committed:
Merge branch 'develop' of github.com:boostcampwm-2022/web37-ZelloDraw into fix/webrtc

HyewonKkang committed:
fix: [FE] 로비,게임페이지에서도 카메라/마이크 세팅

- Intent
로비, 게임 페이지에서도 현재 유저의 카메라와 마이크 상태 세팅이
가능하도록 수정했어요.

HyewonKkang committed:
feat: [FE] camera off인 경우 중앙에 이름 표시

- Intent

HyewonKkang committed:
fix: [FE] 첫 로비 입장 시 remote 유저 캠마이크 세팅

- Intent
서버에서 'webrtc-offer' 이벤트를 보낼 때 유저들의 저장된 audio, video
상태를 전달해요.

LouiMinister commented on frontend/src/pages/Lobby.tsx:
```diff
@@ -41,19 +41,16 @@ function Lobby() {
+                            void createOffers(userInRoom);
+                        }
+                    }, 100);
+                });
```
console.log가 남겨져 있는 의도가 있나요?

HyewonKkang replied:
지우는걸 잊었네요 다시 지웠어요 ㅎㅎ👍🏻

LouiMinister commented on frontend/src/pages/Lobby.tsx:
```diff
@@ -42,12 +42,7 @@ function Lobby() {
-                            console.log('send offer from newbie');
-                            void createOffers(userInRoom);
-                        }
-                    }, 100);
```
setTimeout 도려내기 굿!

HyewonKkang committed:
fix: [FE] 메인페이지 캠, 마이크 토글 버그 해결

- 메인페이지의 비디오 보여주는 컴포넌트를 따로 분리했어요.
- 기존의 useLocalStream()에서 videoRef를 넘겨주던 것이 전역으로 관리하는
  streamRef와 동기화되지 않는 문제가 있어서 ref를 리턴하지 않는 식으로
해결했어요.

- Intent
메인페이지에서 캠, 마이크 on/off가 안되던 오류 해결

HyewonKkang committed:
feat: [FE] 카메라나 마이크가 없는 경우 예외 처리

- Intent
유저의 pc에 카메라나 마이크 디바이스 연결이 안되어 있는 경우에 발생할 수
있는 에러를 처리해주었어요.

HyewonKkang committed:
feat: [FE] 유저 스트림 가져올 때 예외 상황 핸들링

- Intent
RTC 연결 전에 유저 스트림을 가져와야 정상적으로 connection이 생성되는데
스트림을 가져오지 못하는 경우 사용자에게 에러 메세지를 alert로 보여줘요.
추후에 이를 toast message로 변경할 예정이에요.

HyewonKkang committed:
fix: [FE] 디바이스 상태에 따른 요청 보내도록 수정

- Intent
카메나라 마이크 중 하나의 디바이스만 연결된 경우에 그에 맞는 미디어
요청을 보내는 것으로 수정했어요.

HyewonKkang committed:
fix: [FE] 로비 입장 시 보내는 이벤트 순서 수정

- Intent
일부 유저의 audio, video 상태가 'join-lobby'의 response로 오고 있지
않아서 join-lobby를 보내기 전에 update-user-stream emit을 먼저
해주었어요.

HyewonKkang committed:
Merge branch 'develop' into fix/webrtc

HyewonKkang committed:
chore: [FE] @nestjs/websockets package에 추가

- Intent
